Add abstraction layer to create `EXDATE` and `RDATE` properties

axunonb opened the pull request:
### Add abstraction layer to create `EXDATE` and `RDATE` properties
- Introduce `ExceptionDate` and `RecurrenceDates` classes as wrappers for `IList<PeriodList>` properties.
- `Period`s and `CalDateTime`s to them can be done without having to care for RFC 5545 rules regarding serialization:. This happens behind the scenes.

### IRecurrable
- Change the type of `ExceptionDates` from `IList<PeriodList>` to `ExceptionDates`.
- Change the type of `RecurrenceDates` from `IList<PeriodList>` to `RecurrenceDates`.

### RecurringComponent (inhereted by CalendarEvent and other classes), VTimeZoneInfo
- Rename `IList<PeriodList> ExceptionDates` to `ExceptionDatesPeriodLists`. This serves as the backing storage for the new `ExceptionDates` class/property.
- Rename `IList<PeriodList> RecurrenceDates` to `RecurrenceDatesPeriodLists`. This serves as the backing storage for the new `RecurrenceDates` class/property.
- Add the creation of instances of `ExceptionDates` and `RecurrenceDates` to the `Initialize()` method.

### RecurringEvaluator and TodoEvaluator
- Implement the new `ExceptionDates` and `RecurrenceDates` classes.

### VTimeZone
- Enhanced serialization:
  - Before: Each `RDATE` date/time was serialized as a separate `RDATE` propery
  - Now: `RDATE` date/times which can be clustered go into one single `RDATE` property

### Period
- Make parameterless CTOR `internal` to ensure proper initialization by users
- Add internal enum `PeriodKind` to identify the period kind as `Undefined`, `DateTime`, `DateOnly` or `Period` and property `PeriodKind`
- `StartTime`, `EndTime` and `Duration` are immutable
- Timezones and date-only or date-time of `StartTime`and (optional) `EndTime` must be the same
- `CompareTo` uses `AsUtc` for comparing the `StartTime`
- Remove returning a "magic" duration of 1 day, if `EndTime` is null and `StartTime` is date-only. This broke `EXDATE` and `RDATE` of an event, when only a start time exists. Full day is handled by `PeriodSerializer`.
- Added `EffectiveEndTime` and `EffectiveDuration` properties to provide calculated values based on the set values.

### PeriodList
- Change access modifier to `internal`.
- Add `internal PeriodKind PeriodKind`
- Add `internal string? TzId` , returning the TzId of the first period
- `EnsureConsistentTimezoneAndPeriodKind`: The first period determines the `TzId` and `PeriodKind` of the `PeriodList` and all other `Period`s added or inserted must have these properties as well.
- Add `SetService(new PeriodListEvaluator(this))` for `StringReader` CTOR overload
- Add `Add(IDateTime)`
- Remove `public` method `GetGroupedPeriods`
- nullable enable

### PeriodListEvaluator
- Change access modifier to `internal`

### EventEvaluator:
- Update to handle `RDATE` durations correctly

### TodoEvaluator:
- Refactor using `ExceptionDate` and `RecurrenceDates` classes
- Remove method `PeriodWithDuration(Period)` as it became redudant with the refactored `Period` class.

### RecurringEvaluator:
- Refactor using `ExceptionDate` and `RecurrenceDates` classes

### DataTypeSerializer and other serializers, CalendarObjectBase:
- `Activator.CreateInstance(TargetType, true)` allows for not `public` CTORs, so that parameterless CTORs can be excluded from the public API, if proper initialization can't be assured (like with `Period`).

### PeriodListSerializer, PeriodSerializer, DateTimeSerializer, PropertySerializer
- Update serializers to handle the modified `PeriodList` implementation.
- Never set the UTC timezone ID (use appending 'Z') (Section 3.2.19: The "TZID" property parameter MUST NOT be applied to DATE properties and DATE-TIME or TIME properties whose time values are specified in UTC.)
- `RDATE` and `EXDATE` serialize following RFC 5545 sections 3.8.5.2 and 3.8.5.1

Resolves #590
Resolves #591
Resolves #614
Resolves #676

axunonb committed:
Fix `Period` and `PeriodList` to work with `EXDATE` and `RDATE`

Improve reliability and usability for `Period` and `PeriodList` to become less error-prone
by enforcing timezones being used consistently

Period
- Make parameterless CTOR `internal` to ensure proper initialization by users
- A period can be defined
  1. by a start time and an end time,
  2. by a start time and a duration,
  3. by a start time only, with the duration unspecified. This is for EXDATE and RDATE date-only and date/time.
- For cosistency, either the `EndTime` or the `Duration` can be set at a time. The last one set with a value not `null` will prevail, while the other will become `null`.
- Timezones of `StartTime`and (optional) `EndTime` must be the same
- `CompareTo` uses `AsUtc` for comparing the `StartTime`
- Remove returning a "magic" duration of 1 day, if `EndTime` is null and `StartTime` is date-only. This broke `EXDATE` and `RDATE` of an event, when the only a start time exists.
- Added `EffectiveEndTime` and `EffectiveDuration` properties to provide calculated values based on the set values.
- Update the `EndTime` and `Duration` properties to directly return the set values.
- Change the access modifiers of `GetEffectiveDuration()` and `GetEffectiveEndTime` methods from internal to private.

PeriodList
- `TzId`: `public` setter changed to `private`
- `EnsureConsistentTimezones`: The first period determines the timezone of the `PeriodList` and all other `Period`s added must have the same timezone
- Add `SetService(new PeriodListEvaluator(this))` for `StringReader` CTOR overload
- Add `static PeriodList FromStringReader(StringReader)`
- Add `static PeriodList FromDateTime(IDateTime)`
- Add `PeriodList AddPeriod(Period)` for chaining
- Add `PeriodList Add(IDateTime)` for chaining
- nullable enable

EventEvaluator:
- `EventEvaluator.WithEndTime(Period)` only sets the `EndTime`, as `Period.EffectiveDuration` returns the duration.

TodoEvaluator:
- Remove method `PeriodWithDuration(Period)` as it became redudant with the refactored `Period` class.

DataTypeSerializer and other serializers, CalendarObjectBase:
- `Activator.CreateInstance(TargetType, true)` allows for not `public` CTORs, so that parameterless CTORs can be excluded from the public API, if proper initialization can't be assured (like with `Period`).

PropertySerializer:
- TBD: Never set the UTC timezone ID (use appending 'Z')

- Resolves #590
- Resolves #591
- Resolves #614
- Resolves #676

axunonb committed:
Add unit tests to `RecurrenceWithRDateTests`

axunonb committed:
Insert values about failure to Exception.Messages

axunonb committed:
Correct handling of iCalendar PERIOD

PeriodKind
- Add new enum

Period:
- Add internal `property string TzId`
- Add internal method `PeriodKind GetPeriodKind()`
- Ensure timeone consistence for setters of StartTime / EndTime

PeriodList:
- Add internal `property string TzId`
- Add internal property ``PeriodKind PeriodListKind`m
- Ensure added `Period`s have the same `TzId`and `PeriodKind` as the first one added

PeriodListSerialilzer
- Serializes TZID
- Serializes value type PERIOD

axunonb committed:
Add abstraction layer to create EXDATE and RDATE

- Introduced `ExceptionDateCollection`, `RecurrencePeriodCollection` and PeriodCollectionBase` classes.
- Updated serializers to handle the modified `PeriodList` implementation.

These classes are an abstraction layer for creating `List<PeriodList>` objects.
`Period`s and `CalDateTime`s are accepted without having to care for RFC 5545 rules regarding serialization:

`ToRecurrenceDates` and `ToRecurrenceDates` methods aggregate and convert the exception or recurrence `CalDateTime` and `Period` objects into a list of `PeriodList` objects. Periods are grouped by their timezone IDs and period kinds in a way, that each `PeriodList` contains only distinct periods ready to serialize.

minichma commented on Ical.Net/DataTypes/Period.cs:
```diff
@@ -98,51 +115,104 @@ public override string ToString()
     {
-        if (Duration is { } d)
-            return d;
+        if (end?.TzId != null && start.TzId != end.TzId) throw new ArgumentException($"Start time ({start}) and end time ({end}) must have the same timezone.");
```
Here we allow end being floating while start isn't. Should we allow that? I feel we shouldn't because this way we create additional cases that need to be tested and therefore increase overall complexity. I.e. either require the date to be floating, meaning that TzId equals that of Start or require TzId to be set. I recommend the latter (require `left.IsFloating == right.IsFloating`).

axunonb replied:
Absolutely, I had arithmetic in mind here - so be fixed.

minichma commented on Ical.Net/DataTypes/Period.cs:
```diff
@@ -98,51 +115,104 @@ public override string ToString()
+            _endTime = value;
+            if (_endTime != null)
+            {
+                _duration = null;
```
I think we shouldn't allow setting the end time if a duration has already been set. Ideally the whole class would be immutable, but for now I think we should at least not allow changing the structure/kind of an object.

axunonb replied:
Yes, that's what I had in mind as well. When deserializing non-standard ics, this might happen. Disallowing to change the kind: absolutely.

minichma replied:
Well, if we want to support non-standard ICS, then we'd have to allow setting both, but how would we deal with that later on? Probably the deserializer should deal with that case and either raise an error or ignore one of both values.

axunonb replied:
@minichma With the last but one commit StartTime, EndTime and Duration became immutable.
If a Period should change a new instance is the way to. In our code, there are only about 3 places (e.g. recurrence) where this happens.
There is also a new helper method
`internal static Period Create(IDateTime start, IDateTime? end = null, Duration? duration = null, ICalendarObject? associatedObject = null)`
that defines the rule we use, when Duration and EndTime exist. Anyway, there can be only one of them.
If time allows, have a look at the refactored Period class.

minichma commented on Ical.Net/DataTypes/Period.cs:
```diff
@@ -98,51 +115,104 @@ public override string ToString()
+            _duration = value;
+            if (_duration != null)
+            {
+                _endTime = null;
```
See above

> I think we shouldn't allow setting the end time if a duration has already been set. Ideally the whole class would be immutable, but for now I think we should at least not allow changing the structure/kind of an object.

minichma commented on Ical.Net/DataTypes/PeriodList.cs:
```diff
@@ -48,16 +73,18 @@ public override void CopyFrom(ICopyable obj)
+    /// <summary>
+    /// Used for equality comparison of two lists of periods.
+    /// </summary>
+    public static Dictionary<string, IList<Period>> GetGroupedPeriods(IList<PeriodList> periodLists)
```
If I remember correctly, it should be possible to remove this one. Could be a good time now as updating it is not too easy. Should probably be at least made internal.

axunonb replied:
Will go into another (smaller) PR

axunonb commented on Ical.Net.Tests/RecurrenceTests.cs:
```diff
@@ -3201,10 +3201,10 @@ public void OccurrenceMustBeCompletelyContainedWithinSearchRange()
     /// Find a solution for issue #120 or close forever
     /// </summary>
-    [Test, Ignore("No solution for issue #120 yet", Until = "2024-12-31")]
+    [Test, Ignore("No solution for issue #120 yet", Until = "2025-02-28")]
```
As also mentioned in #685, this test is pending since 2016 and should be removed.

axunonb commented on Ical.Net/Calendar.cs:
```diff
@@ -275,20 +275,14 @@ public virtual IEnumerable<Occurrence> GetOccurrences<T>(IDateTime startTime = n
         return default(T);
     }
 
-    public void Dispose()
```
Not in use and and implementing IDisposible

minichma commented on Ical.Net/DataTypes/PeriodList.cs:
```diff
@@ -15,26 +16,50 @@
+    /// </summary>
+    /// <param name="value"></param>
+    /// <exception cref="ArgumentException"></exception>
+    public static PeriodList FromStringReader(StringReader value) => new PeriodList(value);
```
Do we expect this to be commonly used? Maybe internal?

axunonb replied:
Should not be too common. Will remove.

minichma commented on Ical.Net/DataTypes/PeriodList.cs:
```diff
@@ -15,26 +16,50 @@
+    /// </summary>
+    /// <param name="value"></param>
+    /// <exception cref="ArgumentException"></exception>
+    private PeriodList(StringReader value)
```
Do we expect this to be commonly used? Maybe internal?

minichma commented on Ical.Net/DataTypes/Period.cs:
```diff
@@ -98,51 +115,104 @@ public override string ToString()
+        return StartTime.HasTime ? PeriodKind.DateTime : PeriodKind.DateOnly;
+    }
+
+    private Duration GetEffectiveDuration()
```
Shouldn't this be replaced by `EffectiveDuration`? The difference in the current implementation is that the property can return null while the method can't. However, maybe it's easier to deal with that difference at the call site.

minichma commented on Ical.Net/DataTypes/Period.cs:
```diff
@@ -98,51 +115,104 @@ public override string ToString()
     }
 
-    internal IDateTime GetEffectiveEndTime()
+    private IDateTime GetEffectiveEndTime()
```
same as above: Remove in favor of the corresponding property?

axunonb replied:
Yep

minichma commented on Ical.Net/Collections/PeriodCollectionBase.cs:
```diff
@@ -0,0 +1,155 @@
+/// <summary>
+/// Represents a base class for collections of periods.
+/// </summary>
+public abstract class PeriodCollectionBase : ICollection<Period>
```
If we have a `PeriodCollectionBase` shouldn't it be the base class also of `PeriodList`?

Thinking from a different side - the class basically is an implementation of ICollection but doesn't do any additional validation of the entries. Would it make sense to replace the whole class, just like `ExceptionDateCollection` with generic `List<IDateTime>` or the like and implement the specific functionality in extension methods?

axunonb replied:
Good point. I'll try to make `PeriodList` internal as it's only a vehicle for de/serializing. Will have a look.

minichma commented on Ical.Net/Collections/ExceptionDateCollection.cs:
```diff
@@ -0,0 +1,66 @@
+/// each <see cref="PeriodList"/> contains only distinct periods.
+/// </para>
+/// </summary>
+public class ExceptionDateCollection : PeriodCollectionBase
```
Wouldn't it be possible (and cleaner) letting the `ExceptionDateCollection` be a collection of type `IDateTime` rather than `Period`? In case of `RDATE` we need to deal with the different type of entries but in case of EXDATEs this isn't necessary, so it would probably be cleaner to have it strongly typed. Also see my comment in `PeriodCollectionBase`, maybe replace the whole class with just `List<IDateTime>` and implement the specific functionality as extension methods?

minichma replied:
Oh, just realizing that the class is only used in test cases. Shouldn't we move this to the test projects then?

minichma commented on Ical.Net/DataTypes/PeriodList.cs:
```diff
@@ -73,57 +100,116 @@ public static Dictionary<string, List<Period>> GetGroupedPeriods(IList<PeriodLis
+    /// <exception cref="ArgumentException"></exception>
+    public void Add(Period item)
+    {
+        EnsureConsistentTimezoneAndPeriodKind(item);
```
I feel this validation is somewhat hopeless at this time, because there are several other places to verify (Insert, Indexer, CopyTo), but even if we tackle all of them, Period and IDateTime are still mutable, so they could still be changed (another reason why it would be great to have them immuntable).

A question that comes to my mind is whether it would be a viable alternative to allow mixing different kinds altogether, having just a single list in `CalendarEvent` and split into multiple just during serialization? We wouldn't preserve the original structure after a deserialization/serialization cycle, but that should be ok as the semantics would stay unchanged. Would also relieve us from some 'GetGrouped...' kind of code.

minichma commented on Ical.Net/Serialization/DataTypes/PeriodSerializer.cs:
```diff
@@ -34,30 +34,36 @@ public override string SerializeToString(object obj)
+            // "DTEND" nor "DURATION" property, the event’s duration is taken to
+            // be one day:
+
+            if (p.EndTime is { HasTime: true })
```
We need to serialize EndTime also if it is date-only.

axunonb replied:
Fixed already.

minichma reviewed (commented):
This is a challenging one! Left a few comments but they might not be fully consistent. Just take what you think makes sense. Unfortunately I don't have the time right now to go into all the detail that would be required.

axunonb committed:
`Period` StartTime, EndTime and Duration properties are immutable

- Adjust code in ' EventEvaluator` and `PeriodSerializer`
- Method `GetPeriodKind()` is replace with internal property `PeriodKind`
- Add internal `Period.Create` to centralize logic if end time and duration exist
- Modified unit test

axunonb committed:
Remove source and test files related to RDATE/EXDATEcollections

Remove solution for RDATE/ESDATE introduced earlier in this PE

The commit removes several test and source files related to period and exception date collections in calendar events. Specifically, the following files have been completely removed:
- `PeriodCollectionBase.cs`
- `ExceptionDateCollection.cs`
- `RecurrencePeriodCollection.cs`
- `PeriodCollectionTests.cs`
- `RecurrenceWithExDateTests.cs`
- `RecurrenceWithRDateTests.cs`

axunonb committed:
Add wrapper classes for `ExceptionDates` and `RecurrenceDates`

- Add internal properties `PeriodKind` and `TzId` to `PeriodList`.
- Modify `Period` indexer and `Insert` to ensure consistency for timezone and `PeriodKind`
- Prepare for `PeriodList` to become `internal`
- Update `PeriodListSerializer` and `PeriodSerializer` classes for `Period` serialization
- Introduce `ExceptionDates`, `RecurrenceDates`, and `PeriodListWrapperBase` to become wrappers around these properties of `RecurringComponent`
- Add unit tests in `PeriodListWrapperTests` for `ExceptionDates` and `RecurrenceDates`.

Note: The wrappers are not yet implemented for `IRecurringComponent`s.

axunonb committed:
Migrate usage of `IList<PeriodList>` to `ExceptionDates` and `RecurrenceDates`

`IRecurrable`
- Change the type of `ExceptionDates` from `IList<PeriodList>` to `ExceptionDates`.
- Change the type of `RecurrenceDates` from `IList<PeriodList>` to `RecurrenceDates`.

`RecurringComponent`, `VTimeZoneInfo`
- Rename `IList<PeriodList> ExceptionDates` to `ExceptionDatesPeriodLists`. This serves as the backing storage for the new `ExceptionDates` class/property.
- Rename `IList<PeriodList> RecurrenceDates` to `RecurrenceDatesPeriodLists`. This serves as the backing storage for the new `RecurrenceDates` class/property.
- Add the creation of instances of `ExceptionDates` and `RecurrenceDates` to the `Initialize()` method.

`PeriodList` and `PeriodListEvaluator`
- Change the access modifier to `internal`.

`RecurringEvaluator` and `TodoEvaluator`
- Implement the new `ExceptionDates` and `RecurrenceDates` classes.

`VTimeZone`
- Enhanced serialization:
  - Before: Each `RDATE` date/time was serialized as a separate `RDATE` propery
  - Now: `RDATE` date/times which can be clustered go into one single `RDATE` property

Unit tests
- Refactor tests, using `ExceptionDates` and `RecurrenceDates` instead of `IList<PeriodList>`
- Remove the duplicate `SimpleDeserializationTests.RecurrenceDates1()` method, which exists in `DeserializationTests.RecurrenceDates1()`.

Todo:
- Replace the use of `PeriodList.GetGroupedPeriods` in `CalendarEvent.Equals()`.

axunonb committed:
Add `PeriodListWrapperBase.GetAllPeriodsByKind`

Using this internal method stops recreating `Period`s from `IDateTime`
Update comments in PropertySerializer

axunonb committed:
Replace method `PeriodList.GetGroupedPeriods`

- No duplicate `Period` instances can be added or inserted into the `PeriodList`.
- `RecurrenceDates.GetAllDates()`, `PeriodListWrapperBase.GetAllDates()`  and `PeriodListWrapperBase.GetAllPeriodsByKind` return a flattened and ordered list of distinct objects
- Update `CalendarEvent.Equals` and `CalendarEvent.GetHashCode` using `ExceptionDates` and `RecurrenceDates` instead of `PeriodList.GetGroupedPeriods`
- Remove `PeriodList.GetGroupedPeriods`

axunonb committed:
Add EXDATE and RDATE unit tests mentioned in issues

minichma commented on Ical.Net/DataTypes/Period.cs:
```diff
@@ -11,47 +11,97 @@ namespace Ical.Net.DataTypes;
 
-        StartTime = start ?? throw new ArgumentNullException(nameof(start));
-        EndTime = end;
+        _startTime = start ?? throw new ArgumentNullException(nameof(start), "Start time cannot be null.");
```
We'd have seen a NPE if start was null, at least if end was not nulll.

axunonb replied:
Removed

minichma commented on Ical.Net/DataTypes/Period.cs:
```diff
@@ -89,60 +139,81 @@ public override int GetHashCode()
+            {
+                null when _duration is null => null,
+                { } endTime => endTime,
+                _ => EffectiveDuration is not null
```
avoid evaluating EffectiveDuration twice?

axunonb replied:
Changed

minichma commented on Ical.Net/DataTypes/Period.cs:
```diff
@@ -155,14 +226,14 @@ internal IDateTime GetEffectiveEndTime()
+        var endTime = EffectiveEndTime;
         // End time is exclusive
-        return endTime == null || endTime.GreaterThan(dt);
+        return endTime?.GreaterThan(dt) ?? false;
```
This is a functional change, i.e. if EffectiveEndTime is null. Before dt was considered contained and now it isn't. I assume the new version is better, but just to be sure, this was intentional. Could also throw in such a case, because 'Contains' doesn't make much sense in such a case.

axunonb replied:
Must be `return endTime?.GreaterThan(dt) != false;`

minichma commented on Ical.Net/DataTypes/Period.cs:
```diff
@@ -155,14 +226,14 @@ internal IDateTime GetEffectiveEndTime()
     public virtual bool Contains(IDateTime? dt)
```
This kind of convenience methods could be changed to extension methods (in some future PR)

axunonb replied:
agree

minichma commented on Ical.Net/DataTypes/Period.cs:
```diff
@@ -178,21 +249,22 @@ public virtual bool Contains(IDateTime? dt)
         }
 
-        if (StartTime.Equals(other.StartTime))
+        if (StartTime.AsUtc.Equals(other.StartTime.AsUtc))
```
Why do we do a UTC comparison here? I.e. equality is quite a topic in CalDateTime, so why would we use a different comparison here? Do we need the comparison at all? It seems to implement a somewhat special case. I don't think there's a natural order if the duration isn't considered. Maybe remove the comparison altogether and let the caller choose how to compare?

axunonb replied:
Oh yes, you remember the discussion about `CalDateTime` equality.
I opened a [discussion](https://github.com/ical-org/ical.net/discussions/606#discussioncomment-11867080) for that topic.
We can't use Exact Equivalence (same `IsFloating` and same `Value` and same `HasTime` and same `TzId`) here, because the timezone of 2 Periods may be different. and`.AsUtc` is a must.

Isn't the `CompareTo` method specific to each component, that the all callers should "just use".

minichma commented on Ical.Net/DataTypes/PeriodList.cs:
```diff
@@ -3,38 +3,66 @@
+/// for EXDATE and RDATE properties.
 /// </summary>
-public class PeriodList : EncodableDataType, IList<Period>
+internal class PeriodList : EncodableDataType, IList<Period>
```
Like that it's internal now.

minichma commented on Ical.Net/DataTypes/PeriodList.cs:
```diff
@@ -48,82 +76,113 @@ public override void CopyFrom(ICopyable obj)
-
-    public void Add(IDateTime dt) => Periods.Add(new Period(dt));
-
-    public static Dictionary<string, List<Period>> GetGroupedPeriods(IList<PeriodList> periodLists)
```
Great that GetGroupedPeriods is removed!

minichma commented on Ical.Net/DataTypes/PeriodListWrapperBase.cs:
```diff
@@ -0,0 +1,109 @@
+        return periodList;
+    }
+
+    private protected PeriodList? GetPeriodList(IDateTime dt)
```
I don't think you can get a single list for a date. There could be multiple lists having the same TzId and PeriodKind and a single date could be included in more than one list.

minichma commented on Ical.Net/DataTypes/PeriodListWrapperBase.cs:
```diff
@@ -0,0 +1,109 @@
+    /// </summary>
+    public bool Contains(IDateTime dt)
+    {
+        var periodList = GetPeriodList(dt);
```
You get the first list of dt's kind and tzid but not necessarily the one that contains this dt.

minichma commented on Ical.Net/DataTypes/PeriodListWrapperBase.cs:
```diff
@@ -0,0 +1,109 @@
+    /// <inheritdoc cref="PeriodList.Remove"/>
+    public bool Remove(IDateTime dt)
+    {
+        var periodList = GetPeriodList(dt);
```
You get the first list of dt's kind and tzid but not necessarily the one that contains this dt. And dt could be included more than once.

minichma commented on Ical.Net/DataTypes/PeriodListWrapperBase.cs:
```diff
@@ -0,0 +1,109 @@
+    private protected PeriodList? GetPeriodList(Period period)
+    {
+        // The number of PeriodLists is expected to be small, so a linear search is acceptable.
+        return ListOfPeriodList
```
Same as with DateTime, there could be multiple lists holding the same Period.

minichma commented on Ical.Net/DataTypes/PeriodListWrapperBase.cs:
```diff
@@ -0,0 +1,109 @@
+    /// <see cref="PeriodKind.Period"/>, <see cref="PeriodKind.DateOnly"/> and <see cref="PeriodKind.DateTime"/>.
+    /// </summary>
+    internal IEnumerable<Period> GetAllPeriodsByKind(params PeriodKind[] periodKinds)
+        => ListOfPeriodList.SelectMany(pl => pl.Where(p => periodKinds.Contains(p.PeriodKind))).OrderedDistinct();
```
You can use `OrderedDistinct` only if the list it is applied on is ordered. But this would only be the case if the individual lists would be ordered and they would be merged using `OrderedMerge`. But do we need to apply a Distinct at all? If the values are duplicated, why not return them here? The method name suggests that 'all' are returned.

axunonb replied:
Yes - RTD., thanks. When attempting to add a duplicate, it just gets discarded. For consistency, values from deserialization should also be unique.

minichma commented on Ical.Net/DataTypes/RecurrenceDates.cs:
```diff
@@ -0,0 +1,102 @@
+    /// </summary>
+    public bool Contains(Period period)
+    {
+        var periodList = GetPeriodList(period);
```
See above - GetPriodList doesn't necessarily return the right list.

minichma commented on Ical.Net/DataTypes/RecurrenceDates.cs:
```diff
@@ -0,0 +1,102 @@
+    /// <inheritdoc cref="PeriodList.Remove"/>
+    public bool Remove(Period period)
+    {
+        var periodList = GetPeriodList(period);
```
see above, possibly not the right list

minichma commented on Ical.Net/CalendarComponents/CalendarEvent.cs:
```diff
@@ -138,15 +138,13 @@ calculation from the zoned start time to the zoned end time.
+        // specifies a "DTSTART" property with a DATE value type but no
+        // "DTEND" nor "DURATION" property, the event’s duration is taken to
+        // be one day.
+        // This is taken care of in the PeriodSerializer class,
```
Don't think we should remove this as the Period only comes into play during evaluation, but the CalendarEvent's duration could be queried independently. Either we remove the method altogether or we return a precise value. We could consolidate the calculation though. I would also not consider this a 'magic number' though. It's just the value that's specified in the RFC.

We have the same kind of logic in multiple places now, so we could consolidate it in a utility class.

axunonb replied:
Valid point - reverted

minichma commented on Ical.Net/CalendarComponents/CalendarEvent.cs:
```diff
@@ -287,29 +285,18 @@ protected bool Equals(CalendarEvent? other)
-        var otherRDates = PeriodList.GetGroupedPeriods(other.RecurrenceDates);
-        if (rDates.Keys.Count != otherRDates.Keys.Count || !rDates.Keys.OrderBy(k => k).SequenceEqual(otherRDates.Keys.OrderBy(k => k)))
+        // exDates and otherExDates are filled with a sorted list of distinct periods
+        var exDates = ExceptionDates.GetAllPeriodsByKind(PeriodKind.Period, PeriodKind.DateOnly, PeriodKind.DateTime).ToList();
```
Do we need an implementation of Equals at all? Not sure.

axunonb replied:
I guess, we don't for `CalendarEvent`. Should go into another PR.

minichma reviewed (commented):
@axunonb Had a rather brief look to the most relevant portions and left a few comments again, maybe some duplicated. Please see, what you can use.

axunonb committed:
Updates after revview 2025-01-16
